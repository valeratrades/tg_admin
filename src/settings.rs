use anyhow::Result;
use config::Config;
use v_utils::macros::MyConfigPrimitives;

#[derive(Clone, Debug, Default, MyConfigPrimitives)]
pub struct Settings {
	pub tg_token: String,
}

impl Settings {
	/// Uses config at the provided position and considers the cli arguments. Cli overrides the config.
	pub fn new_with_cli(cli: &crate::Cli) -> Result<Self> {
		let crate::Commands::Start(cmd_args) = &cli.command;

		let s = Config::builder()
			.add_source(config::File::with_name(&cli.config.to_string()).required(false))
			.set_override_option("tg_token", cmd_args.tg_token.clone())?
			.build()?;

		let settings: Self = s.try_deserialize()?;

		Ok(settings)
	}
}
