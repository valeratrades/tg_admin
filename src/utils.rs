use serde_json::Value;

pub fn get_json_type(value: &Value) -> &str {
	match value {
		Value::Null => "Null",
		Value::Bool(_) => "Boolean",
		Value::Number(_) => "Number",
		Value::String(_) => "String",
		Value::Array(_) => "Array",
		Value::Object(_) => "Object",
	}
}

pub fn value_preview(key: &String, value: &Value) -> String {
	match value {
		Value::Object(_) => format!("{{}} {}", key),
		Value::Array(arr) => format!("[{}] {}", arr.len(), key),
		_ => format!("{}: {}", key, &value.to_string()),
	}
}
