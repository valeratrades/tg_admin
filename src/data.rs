use crate::utils::get_json_type;
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use serde_yaml::Value as YamlValue;
use std::io::{Read, Write};
use std::{
	fs::File,
	io::{BufReader, BufWriter},
	path::{Path, PathBuf},
};
use toml::Value as TomlValue;

#[derive(Clone, Debug, Default, derive_new::new)]
pub struct Data {
	inner: JsonValue,
	path: PathBuf,
}

impl Data {
	/// Load data from a file
	pub fn load(path: &Path) -> Result<Self> {
		let file = File::open(path)?;
		let mut reader = BufReader::new(file);
		let extension = path.extension().and_then(std::ffi::OsStr::to_str).unwrap_or("");

		let data: JsonValue = match extension {
			"json" | "json5" => {
				let mut content = String::new();
				reader.read_to_string(&mut content)?;
				let json5_value: JsonValue = json5::from_str(&content).context("Failed to read JSON file")?;
				json5_value
			}
			"yaml" | "yml" => {
				let yaml_value: YamlValue = serde_yaml::from_reader(reader).context("Failed to read YAML file")?;
				serde_json::to_value(yaml_value).context("Failed to convert YAML to JSON")?
			}
			"toml" => {
				let mut content = String::new();
				reader.into_inner().read_to_string(&mut content)?;
				let toml_value: TomlValue = toml::from_str(&content).context("Failed to read TOML file")?;
				serde_json::to_value(toml_value).context("Failed to convert TOML to JSON")?
			}
			_ => return Err(anyhow::anyhow!("Unsupported file format")),
		};

		Ok(Self::new(data, path.to_path_buf()))
	}

	/// Write data to the source file
	pub fn write(&self) -> Result<()> {
		let file = File::create(&self.path)?;
		let mut writer = BufWriter::new(file);
		let extension = self.path.extension().and_then(std::ffi::OsStr::to_str).unwrap_or("");

		match extension {
			"json" | "json5" => serde_json::to_writer(writer, &self.inner).context("Failed to write JSON file")?,
			"yaml" | "yml" => {
				let yaml_value: YamlValue = serde_json::from_value(self.inner.clone()).context("Failed to convert JSON to YAML")?;
				serde_yaml::to_writer(writer, &yaml_value).context("Failed to write YAML file")?
			}
			"toml" => {
				let toml_value: TomlValue = serde_json::from_value(self.inner.clone()).context("Failed to convert JSON to TOML")?;
				writer.write_all(toml::to_string(&toml_value).context("Failed to write TOML file")?.as_bytes())?;
			}
			_ => return Err(anyhow::anyhow!("Unsupported file format")),
		}

		Ok(())
	}

	/// Load the file without needing to provide the path again
	pub fn reload(&mut self) -> Result<()> {
		self.inner = Self::load(&self.path)?.inner;
		Ok(())
	}

	pub fn at(&self, level: &ValuePath) -> Option<JsonValue> {
		let mut current = &self.inner;
		for part in level.to_vec() {
			current = current.get(&part)?;
		}
		Some(current.clone())
	}

	pub fn update(&mut self, new_value: JsonValue) {
		self.inner = new_value;
	}

	pub fn update_at<UA>(&mut self, level: &ValuePath, new_value: JsonValue, into_action: UA) -> Result<(), String>
	where
		UA: Into<UpdateAction>,
	{
		let mut current = &mut self.inner;
		let path = level.to_vec();
		let action = into_action.into();

		for (i, part) in path.iter().enumerate() {
			if i == path.len() - 1 {
				// We're at the last element, so we apply the update action
				let obj = current.as_object_mut().unwrap();
				match action {
					UpdateAction::Set => {
						obj.insert(part.clone(), new_value);
					}
					UpdateAction::AddTo => {
						let existing = obj.get_mut(part).unwrap();
						if let JsonValue::Array(existing_arr) = existing {
							if !existing_arr.is_empty() && get_json_type(&existing_arr[0]) != get_json_type(&new_value) {
								return Err(format!(
									"Type mismatch: Expected {}, got {}",
									get_json_type(&existing_arr[0]),
									get_json_type(&new_value)
								));
							}
							existing_arr.push(new_value);
						} else {
							panic!("Target is not an array");
						}
					}
					UpdateAction::RemoveFrom => {
						dbg!(&part);
						let existing = obj.get_mut(part).unwrap();
						if let JsonValue::Array(existing_arr) = existing {
							if existing_arr.is_empty() {
								return Err("Cannot remove from an empty array".to_string());
							}
							if get_json_type(&existing_arr[0]) != get_json_type(&new_value) {
								return Err(format!(
									"Type mismatch: Expected {}, got {}",
									get_json_type(&existing_arr[0]),
									get_json_type(&new_value)
								));
							}
							let initial_len = existing_arr.len();
							existing_arr.retain(|item| item != &new_value);
							if existing_arr.len() == initial_len {
								return Err("Value not found in the array".to_string());
							}
						} else {
							panic!("Target is not an array");
						}
					}
				}
				return Ok(());
			} else {
				// We're not at the last element, so we navigate to the next level
				current = current.get_mut(part).unwrap();
			}
		}
		panic!("Empty path");
	}

	#[doc(hidden)]
	pub fn mock(value: JsonValue) -> Self {
		Self::new(value, PathBuf::new())
	}
}
impl AsRef<JsonValue> for Data {
	fn as_ref(&self) -> &JsonValue {
		&self.inner
	}
}
#[derive(Clone, Debug, PartialEq, Eq, Copy)]
pub enum UpdateAction {
	Set,
	AddTo,
	RemoveFrom,
}
impl From<crate::telegram::InputValueType> for UpdateAction {
	fn from(action: crate::telegram::InputValueType) -> Self {
		match action {
			crate::telegram::InputValueType::UpdateAt => Self::Set,
			crate::telegram::InputValueType::AddTo => Self::AddTo,
			crate::telegram::InputValueType::RemoveFrom => Self::RemoveFrom,
		}
	}
}

/// Callback data must never be empty, so 0 level is "/" and not ""
#[derive(Clone, Debug, derive_new::new, Serialize, Deserialize, PartialEq, Eq)]
pub struct ValuePath(String);
impl ValuePath {
	pub fn push(&mut self, part: &str) {
		if !self.is_top() {
			self.0.push('/');
		}
		self.0.push_str(part);
	}

	pub fn parent(&self) -> Self {
		let mut new_v = self.0.clone();
		let pos = new_v.rfind("/").unwrap_or(0);
		new_v.truncate(pos);
		match new_v.is_empty() {
			true => Self::default(),
			false => Self(new_v),
		}
	}

	pub fn basename(&self) -> String {
		let pos = self.0.rfind("/").unwrap_or(0);
		let offset = Self::default().0.len();
		self.0[(pos + offset)..].to_string()
	}

	pub fn join(&self, part: &str) -> Self {
		let mut new_level = self.clone();
		new_level.push(part);
		new_level
	}

	pub fn is_top(&self) -> bool {
		assert!(!self.0.is_empty());
		self.0 == "/"
	}

	fn to_vec(&self) -> Vec<String> {
		self.0.split("/").map(String::from).filter(|v| v != "").collect()
	}

	pub fn into_string(self) -> String {
		self.0
	}
}
impl Default for ValuePath {
	fn default() -> Self {
		Self("/".to_string())
	}
}
impl From<Vec<String>> for ValuePath {
	fn from(parts: Vec<String>) -> Self {
		let s = "/".to_owned() + &parts.join("/");
		Self(s)
	}
}
impl From<&str> for ValuePath {
	fn from(s: &str) -> Self {
		Self(s.to_string())
	}
}
impl From<String> for ValuePath {
	fn from(s: String) -> Self {
		Self(s)
	}
}
impl From<ValuePath> for Vec<String> {
	fn from(level: ValuePath) -> Self {
		level.to_vec()
	}
}
impl std::fmt::Display for ValuePath {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(f, "{}", self.0)
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use serde_json::json;
	use std::fs::{create_dir_all, write};
	use tempfile::tempdir;

	fn generate_test_data(format: &str) -> (PathBuf, String) {
		let dir = tempdir().unwrap();
		let path = dir.path().join(format!("test_data.{}", format));

		let json_value = json!({
			"key": "value",
			"number": 42,
			"array_of_numbers": [1, 2, 3],
			"array_of_strings": ["a", "b", "c"]
		});

		let content = match format {
			"json" => serde_json::to_string_pretty(&json_value).unwrap(),
			"yaml" | "yml" => serde_yaml::to_string(&json_value).unwrap(),
			"toml" => {
				let toml_value: toml::Value = serde_json::from_value(json_value).unwrap();
				toml::to_string(&toml_value).unwrap()
			}
			_ => panic!("Unsupported format"),
		};

		(path.to_path_buf(), content)
	}

	#[test]
	fn test_data_operations() -> Result<()> {
		let formats = vec!["json", "yaml", "yml", "toml"];
		for format in formats {
			let (path, content) = generate_test_data(format);

			let dir = path.parent().unwrap();
			create_dir_all(dir)?;

			write(&path, content)?;

			let mut data = Data::load(&path)?;
			assert_eq!(data.as_ref()["key"], "value");
			assert_eq!(data.as_ref()["number"], 42);

			let mut new_inner = data.as_ref().clone();
			if let Some(obj) = new_inner.as_object_mut() {
				obj.insert("key".to_string(), JsonValue::String("new_value".to_string()));
			}
			data.update(new_inner);
			data.write()?;

			let data = Data::load(&path)?;
			assert_eq!(data.as_ref()["key"], "new_value", "(Format: {})", format);
			assert_eq!(data.as_ref()["number"], 42, "(Format: {})", format);
		}
		Ok(())
	}

	#[test]
	fn test_data_update_at() {
		let formats = vec!["json", "yaml", "yml", "toml"];
		for format in formats {
			let (path, content) = generate_test_data(format);

			let dir = path.parent().unwrap();
			create_dir_all(dir).unwrap();

			write(&path, content).unwrap();

			let mut data = Data::load(&path).unwrap();

			// Test UpdateAction::Set
			assert!(data
				.update_at(&ValuePath::from("key"), JsonValue::String("updated_value".to_string()), UpdateAction::Set)
				.is_ok());
			assert_eq!(data.as_ref()["key"], "updated_value", "(Format: {})", format);

			// Test UpdateAction::AddTo for numbers
			let numbers_path = ValuePath::from("array_of_numbers");
			assert!(data.update_at(&numbers_path, JsonValue::Number(4.into()), UpdateAction::AddTo).is_ok());
			assert_eq!(
				data.as_ref()["array_of_numbers"],
				JsonValue::Array(vec![1, 2, 3, 4].into_iter().map(|n| JsonValue::Number(n.into())).collect()),
				"(Format: {})",
				format
			);

			// Test UpdateAction::AddTo for strings
			let strings_path = ValuePath::from("array_of_strings");
			assert!(data.update_at(&strings_path, JsonValue::String("d".to_string()), UpdateAction::AddTo).is_ok());
			assert_eq!(
				data.as_ref()["array_of_strings"],
				JsonValue::Array(vec!["a", "b", "c", "d"].into_iter().map(|s| JsonValue::String(s.to_string())).collect()),
				"(Format: {})",
				format
			);

			// Test UpdateAction::RemoveFrom for numbers
			assert!(data.update_at(&numbers_path, JsonValue::Number(2.into()), UpdateAction::RemoveFrom).is_ok());
			assert_eq!(
				data.as_ref()["array_of_numbers"],
				JsonValue::Array(vec![1, 3, 4].into_iter().map(|n| JsonValue::Number(n.into())).collect()),
				"(Format: {})",
				format
			);

			// Test error cases
			assert!(data
				.update_at(&numbers_path, JsonValue::String("not a number".to_string()), UpdateAction::AddTo)
				.is_err());
			assert!(data.update_at(&numbers_path, JsonValue::Number(5.into()), UpdateAction::RemoveFrom).is_err());
			data.write().unwrap();

			// Verify persistence
			let reloaded_data = Data::load(&path).unwrap();
			assert_eq!(reloaded_data.as_ref()["key"], "updated_value", "(Format: {})", format);
			assert_eq!(
				reloaded_data.as_ref()["array_of_numbers"],
				JsonValue::Array(vec![1, 3, 4].into_iter().map(|n| JsonValue::Number(n.into())).collect()),
				"(Format: {})",
				format
			);
			assert_eq!(
				reloaded_data.as_ref()["array_of_strings"],
				JsonValue::Array(vec!["a", "b", "c", "d"].into_iter().map(|s| JsonValue::String(s.to_string())).collect()),
				"(Format: {})",
				format
			);
		}
	}

	#[test]
	fn test_value_path() {
		let mut level = ValuePath::default();
		let path = ["key1", "key2", "key3"];

		assert!(level.0 != "");
		assert!(level.basename() == "");
		assert!(!level.to_vec().contains(&"".to_string()));

		for part in &path {
			level.push(part);
			assert!(level.parent().0 != "");
			assert!(!level.to_vec().contains(&"".to_string()));
			assert!(level.basename() == *part);
		}
		assert!(level.to_vec() == path.to_vec());
	}
}
