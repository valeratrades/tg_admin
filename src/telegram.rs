use crate::data::{Data, ValuePath};
use crate::settings::Settings;
use anyhow::Result;
//use dptree::HandlerResult;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::{Arc, RwLock};
use teloxide::dispatching::dialogue::{self, InMemStorage};
use teloxide::dispatching::UpdateHandler;
use teloxide::prelude::*;
use teloxide::types::{CallbackQuery, InlineKeyboardButton, InlineKeyboardMarkup, Message, MessageId};
use teloxide::utils::command::BotCommands;
use tracing::info;
use crate::utils::{get_json_type, value_preview};

type MyDialogue = Dialogue<ChatState, InMemStorage<ChatState>>;
type HandlerResult = Result<(), Box<dyn std::error::Error + Send + Sync>>; //dbg
#[derive(Clone, Debug, Default, derive_new::new, Serialize, Deserialize, PartialEq, Eq)]
struct NavigationMessageId(i32);


#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
enum ChatState {
	/// Most actions are prohibited from this state. Other states can be reached only through authorization from here.
	#[default]
	Unauthorized,
	/// Dummy state, only here to not have to spawn Navigation with random value on authorization.
	Authorized,
	Navigation { message_id: i32 },
	Input(ValueInput),
}
#[derive(Clone, Debug, derive_new::new, Serialize, Deserialize, PartialEq, Eq)]
struct ValueInput {
	input_type: InputValueType,
	value_path: ValuePath,
}
#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum InputValueType {
	UpdateAt,
	AddTo,
	RemoveFrom,
}
#[derive(BotCommands, Clone, Debug)]
#[command(description = "Commands:", rename_rule = "lowercase")]
enum Command {
	#[command(description = "Display all commands")]
	Help,
	#[command(description = "Open admin panel at the top")]
	Admin,
}

#[tracing::instrument]
pub async fn run(settings: Arc<Settings>, data: Arc<RwLock<Data>>) -> Result<()> {
	let token = &settings.tg_token;
	let bot = Bot::new(token);
	info!("Starting telegram bot...");
	Dispatcher::builder(bot, schema())
		.dependencies(dptree::deps![data, settings, InMemStorage::<ChatState>::new()])
		.error_handler(LoggingErrorHandler::with_custom_text("An error has occurred in the dispatcher"))
		.enable_ctrlc_handler()
		.build()
		.dispatch()
	.await;
	Ok(())
}

fn schema() -> UpdateHandler<Box<dyn std::error::Error + Send + Sync + 'static>> {
	use dptree::case;

	let command_handler = teloxide::filter_command::<Command, _>()
		.branch(case![Command::Help].endpoint(help_handler))
		.branch(case![Command::Admin].endpoint(admin_handler));

	let message_handler = Update::filter_message()
		.branch(command_handler)
		.branch(case![ChatState::Input(value_input)].endpoint(value_input_handler))
		.branch(dptree::endpoint(invalid_state_handler));

	let callback_query_handler = Update::filter_callback_query().endpoint(callback_query_handler);

	let auth_handler = dptree::filter_map_async(|dialogue: MyDialogue, settings: Arc<Settings>, update: Update| async move {
		match dialogue.get().await {
			Ok(Some(ChatState::Unauthorized)) => {
				if let Some(admin_list) = &settings.admin_list {
					let user_id = update.user().unwrap().id.0;
					if !admin_list.contains(&user_id) {
						return None;  // Not authorized
					}
				}
				dialogue.update(ChatState::Authorized).await.ok()?;
				Some(())  // Authorized
			},
			Ok(Some(_)) => Some(()),  // Already authorized
			_ => None,  // Error or no state, treat as unauthorized
		}
	});

	dialogue::enter::<Update, InMemStorage<ChatState>, ChatState, _>()
		.chain(auth_handler)
		.branch(message_handler)
		.branch(callback_query_handler)
}

async fn admin_handler(bot: Bot, msg: Message, dialogue: MyDialogue, data: Arc<RwLock<Data>>) -> HandlerResult {
	let value_path = ValuePath::default();
	let (header, markup) = {
		let data = data.read().unwrap();
		render_header_and_markup(&data, &value_path)

	};
	let sent_message = bot.send_message(msg.chat.id, &header)
		.reply_markup(markup)
	.await?;
	dialogue.update(ChatState::Navigation { message_id: sent_message.id.0 }).await?;
	Ok(())
}

async fn value_input_handler(
	bot: Bot,
	dialogue: MyDialogue,
	msg: Message,
	value_input: ValueInput,
	data: Arc<RwLock<Data>>,
) -> HandlerResult {
	match msg.text().map(ToOwned::to_owned) {
		Some(new_value) => {
			if let Ok(new_value) = serde_json::from_str::<Value>(&new_value) {
				let update_result = {
					let mut data_lock = data.write().unwrap();
					let result = data_lock.update_at(&value_input.value_path, new_value.clone(), value_input.input_type);
					data_lock.write().unwrap();
					result
				};

				match update_result {
					Ok(_) => {
						let affirmation_menu = match value_input.input_type {
							InputValueType::UpdateAt => {
								format!(
									"Value of `{}` has been updated to `{}`",
									&value_input.value_path,
									&new_value.to_string()
								)
							}
							InputValueType::AddTo => {
								format!(
									"`{}` has been added to `{}`",
									&new_value.to_string(),
									&value_input.value_path
								)
							}
							InputValueType::RemoveFrom => {
								format!(
									"`{}` has been removed from `{}`",
									&new_value.to_string(),
									&value_input.value_path
								)
							}
						};
						bot.send_message(msg.chat.id, affirmation_menu).await?;

						// Resend the nav menu
						let (header, markup) = {
							let data = data.read().unwrap();
							let new_path = match value_input.input_type {
								InputValueType::UpdateAt => value_input.value_path.parent(),
								InputValueType::AddTo | InputValueType::RemoveFrom => value_input.value_path,
							};
							render_header_and_markup(&data, &new_path)
						};
						let sent_message = bot
							.send_message(dialogue.chat_id(), &header)
							.reply_markup(markup)
						.await?;
						dialogue
							.update(ChatState::Navigation {
								message_id: sent_message.id.0,
						})
						.await?;
					}
				Err(e) => {
						bot.send_message(
							msg.chat.id,
							e,
						)
						.await?;
					}
				}
			} else {
				bot.send_message(msg.chat.id, "Invalid value. Input valid JSON value.")
				.await?;
			}
		}
		None => {
			bot.send_message(msg.chat.id, "Please send the new value.")
			.await?;
		}
	}
	Ok(())
}

async fn invalid_state_handler(bot: Bot, msg: Message) -> HandlerResult {
	bot.send_message(msg.chat.id, "Unable to handle the message. Type /help to see available commands.")
	.await?;
	Ok(())
}
async fn help_handler(bot: Bot, msg: Message) -> HandlerResult {
	bot.send_message(msg.chat.id, Command::descriptions().to_string()).await?;
	Ok(())
}

async fn callback_query_handler(bot: Bot, dialogue: MyDialogue, q: CallbackQuery, data: Arc<RwLock<Data>>) -> HandlerResult {
	bot.answer_callback_query(q.id.clone()).await?; // normally this is done after, but I like how it stops for a moment before the action is performed. Otherwise looks cut.
	if let Some(j) = q.data {
		let action: CallbackAction = serde_json::from_str(&j).unwrap();
		match action {
			CallbackAction::Go(value_path) => {
				continue_navigation(bot.clone(), dialogue, data, value_path).await?;
			},
			CallbackAction::UpdateAt(value_path) => {
				dialogue.update(ChatState::Input(ValueInput::new(InputValueType::UpdateAt, value_path.clone()))).await?;
				bot.send_message(
					dialogue.chat_id(),
					format!(
						"You're updating `{}: {}`.\n Insert the new value.",
						&value_path.basename(),
						{
							let data_lock = data.read().unwrap();
							get_json_type(&data_lock.at(&value_path).unwrap())
						}
					),
				).await?;
			},
			CallbackAction::AddTo(value_path) => {
				dialogue.update(ChatState::Input(ValueInput::new(InputValueType::AddTo, value_path.clone()))).await?;
				bot.send_message(
					dialogue.chat_id(),
					format!(
						"You're adding to {}.\n Provide the value to add.",
						value_path
					),
				).await?;
			},
			CallbackAction::RemoveFrom(value_path) => {
				dialogue.update(ChatState::Input(ValueInput::new(InputValueType::RemoveFrom, value_path.clone()))).await?;
				bot.send_message(
					dialogue.chat_id(),
					format!(
						"You're removing from {}.\n Provide exact value to remove.",
						value_path
					),
				).await?;
			},
		}
	}
	Ok(())
}

async fn continue_navigation(bot: Bot, dialogue: MyDialogue, data: Arc<RwLock<Data>>, value_path: ValuePath) -> HandlerResult {
	let (header, markup) = {
		let data = data.read().unwrap();
		render_header_and_markup(&data, &value_path)
	};

	let state = dialogue.get().await.unwrap().unwrap();
	let message_id = match state {
		ChatState::Navigation { message_id } => message_id,
		_ => unreachable!(),
	};

	match bot.edit_message_text(dialogue.chat_id(), MessageId(message_id), &header)
		.reply_markup(markup.clone())
	.await
	{
		Ok(_) => Ok(()),
		//TODO!: assert that the err is about message being too old, as it's the only recoverable one.
		Err(err) => {
			dbg!(err);
			let sent_message = bot.send_message(dialogue.chat_id(), &header)
				.reply_markup(markup)
			.await?;
			dialogue.update(ChatState::Navigation { message_id: sent_message.id.0 }).await?;
			Ok(())
		}
	}
}

#[derive(Clone, Debug, Default, derive_new::new, Serialize, Deserialize, PartialEq, Eq)]
struct More {
	pub value_path: ValuePath,
	pub start_at: usize,
}
#[derive(Clone, Debug, derive_new::new, Serialize, Deserialize, PartialEq, Eq)]
enum CallbackAction {
	Go(ValuePath),
	UpdateAt(ValuePath),
	AddTo(ValuePath),
	RemoveFrom(ValuePath),
}

fn render_header_and_markup(data: &Data, value_path: &ValuePath) -> (String, InlineKeyboardMarkup) {
	let mut keyboard = Vec::new();
	let current_value_at_path = &data.at(value_path).unwrap();
	let mut header = value_path.to_string();

	// Add parent navigation button if not at top level
	if !value_path.is_top() {
		let callback_action = CallbackAction::Go(value_path.parent());
		let button = InlineKeyboardButton::callback("..", serde_json::to_string(&callback_action).unwrap());
		keyboard.push(vec![button]);
	}

	match current_value_at_path {
		Value::Object(map) => {
			for (key, val) in map {
				let (display_text, callback_data) = match val {
					Value::Object(_) | Value::Array(_) => {
						(value_preview(key, val), CallbackAction::Go(value_path.join(key)))
					},
					_ => (value_preview(key, val), CallbackAction::UpdateAt(value_path.join(key))),
				};

				let button = InlineKeyboardButton::callback(display_text, serde_json::to_string(&callback_data).unwrap());
				keyboard.push(vec![button]);
			}
		},
		Value::Array(arr) => {
			header.push_str(&format!(" [{}]", arr.len()));

			let start = arr.len().saturating_sub(25);
			let mut array_str = "\n```json\n".to_owned();
			for a in arr.iter().skip(start) {
				array_str.push_str(&format!("{a}\n"));
			}
			array_str.push_str("```");
			header += &array_str;

			let bottom_row = vec![
				InlineKeyboardButton::callback("Add", serde_json::to_string(&CallbackAction::AddTo(value_path.clone())).unwrap()),
				InlineKeyboardButton::callback("Remove", serde_json::to_string(&CallbackAction::RemoveFrom(value_path.clone())).unwrap()),
			];
			//TODO!: make doubled horizontally `<-` and `->` buttons that modify starting position of the count
			keyboard.push(bottom_row);
		},
		_ => {
			unreachable!();
		}
	}

	(header, InlineKeyboardMarkup::new(keyboard))
}

#[cfg(test)]
mod tests {
	use super::*;
	use serde_json::json;

	fn gen_data() -> (Data, ValuePath) {
		let json_value = json!({
			"name": "Alice",
			"age": 25,
			"address": {
			"street": "456 Another St",
			"city": "Elsewhere"
			},
			"emails": ["alice@example.com", "a@example.com"]
		});
		(Data::mock(json_value), ValuePath::default())
	}

	#[test]
	fn test_top_value_path_representation() {
		let (data, value_path) = gen_data();
		let (_h, r) = render_header_and_markup(&data, &value_path);

		insta::assert_json_snapshot!(
			r,
			@r###"
  {
    "inline_keyboard": [
      [
        {
          "text": "{} address",
          "callback_data": "{\"Go\":\"/address\"}"
        }
      ],
      [
        {
          "text": "age: 25",
          "callback_data": "{\"UpdateAt\":\"/age\"}"
        }
      ],
      [
        {
          "text": "[2] emails",
          "callback_data": "{\"Go\":\"/emails\"}"
        }
      ],
      [
        {
          "text": "name: \"Alice\"",
          "callback_data": "{\"UpdateAt\":\"/name\"}"
        }
      ]
    ]
  }
  "###
		);
	}

	#[test]
	fn test_nested_value_path_representation() {
		let (data, mut value_path) = gen_data();
		value_path.push("address");
		let (_h, r) = render_header_and_markup(&data, &value_path);
		insta::assert_json_snapshot!(
			r,
			@r###"
  {
    "inline_keyboard": [
      [
        {
          "text": "..",
          "callback_data": "{\"Go\":\"/\"}"
        }
      ],
      [
        {
          "text": "city: \"Elsewhere\"",
          "callback_data": "{\"UpdateAt\":\"/address/city\"}"
        }
      ],
      [
        {
          "text": "street: \"456 Another St\"",
          "callback_data": "{\"UpdateAt\":\"/address/street\"}"
        }
      ]
    ]
  }
  "###
		);
	}
	#[test]
	fn test_array_value_path_representation() {
		let (data, mut value_path) = gen_data();
		value_path.push("emails");
		let (h, r) = render_header_and_markup(&data, &value_path);

		insta::assert_snapshot!(
			h,
			"Admin Menu",
		);

		insta::assert_json_snapshot!(
			r,
			@r###"
  {
    "inline_keyboard": [
      [
        {
          "text": "..",
          "callback_data": "{\"Go\":\"/\"}"
        }
      ],
      [
        {
          "text": "Add",
          "callback_data": "{\"AddTo\":\"/emails\"}"
        },
        {
          "text": "Remove",
          "callback_data": "{\"RemoveFrom\":\"/emails\"}"
        }
      ]
    ]
  }
  "###
		);
	}
}
