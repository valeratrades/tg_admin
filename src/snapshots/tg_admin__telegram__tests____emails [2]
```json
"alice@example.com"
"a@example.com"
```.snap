---
source: src/telegram.rs
expression: "\"Admin Menu\""
---
Admin Menu
